### PR TITLE
Create Kafka Producer and Consumer

### PR DESCRIPTION
**Pull request by leowei01**

### CODE REVIEW

File: src/kafka/consumer.clj [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+(comment
+
+  ;; generate config
+  (def consumer-config (new Properties))
+  (.putAll consumer-config {"group.id"           "My-Group"
+                            "key.deserializer"   StringDeserializer
+                            "value.deserializer" StringDeserializer
+                            "client.id"          (.getHostName (InetAddress/getLocalHost))
+                            "bootstrap.servers"  "localhost:9092"})
```
Comment (lukaszhall) [inline src/kafka/consumer.clj]: Move this into the codebase

File: src/kafka/consumer.clj [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+  ;; create a customer client
+  (def consumer (atom (build-consumer consumer-config "demo")))
```
Comment (lukaszhall) [inline src/kafka/consumer.clj]: atoms do _not_ live in the RCB

File: src/kafka/producer.clj [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+            "acks"              "all"})
+
+  ;; create a producer client
+  (def producer (atom (build-producer producer-config)))
+
+  ;; generate and send a record
+  (def record (new ProducerRecord "demo" "Hello world!"))
+  (swap! producer send-msg! record)
+
+  )
```
Comment (lukaszhall) [inline src/kafka/producer.clj]: Same comments as for consumer

Comment (lukaszhall) [reply src/kafka/producer.clj]: Combine these two namespaces

File: src/kafka/consumer.clj [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+  ;; while loop to receive message
+  (swap! consumer read-msg!)
```
Comment (lukaszhall) [inline src/kafka/consumer.clj]: misuse of swap!.
This is NOT an idempotent function. Read the documentation for swap!

File: src/kafka/kafka_clients.clj [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+           (org.apache.kafka.common.serialization StringSerializer StringDeserializer)
+           (org.apache.kafka.clients.consumer KafkaConsumer)
+           (java.time Duration)))
+
+(def producer-config (new Properties))
+(.putAll producer-config {"value.serializer"  StringSerializer
+                          "key.serializer"    StringSerializer
+                          "client.id"         (.getHostName (InetAddress/getLocalHost))
+                          "bootstrap.servers" "localhost:9092"
+                          "acks"              "all"})
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj]: dangerous

File: src/kafka/kafka_clients.clj [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+(def consumer-config (new Properties))
+(.putAll consumer-config {"group.id"           "My-Group"
+                          "key.deserializer"   StringDeserializer
+                          "value.deserializer" StringDeserializer
+                          "client.id"          (.getHostName (InetAddress/getLocalHost))
+                          "bootstrap.servers"  "localhost:9092"})
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj]: dangerous

Comment (lukaszhall) [reply src/kafka/kafka_clients.clj]: These should each have fns for building the relative configs

File: src/kafka/kafka_clients.clj [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+  [producer record]
+  (.send producer record))
+
+(defn read-msg!
+  [consumer]
+  (while true
+    (let [records (.poll consumer (Duration/ofMillis 100))]
+      (doseq [record records]
+        (println (str "Processed Value: " (.value record)))))
+    (.commitAsync consumer)))
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj]: This is inconsistent with our data oriented approach. Why wouldn't this fn return the messages?

File: src/kafka/kafka_clients.clj [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+  [config]
+  (new KafkaProducer config))
+
+(defn build-consumer
+  [config topic]
+  (let [consumer (new KafkaConsumer config)]
+    (.subscribe consumer [topic])
+    consumer))
+
+(def producer (build-producer producer-config))
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj]: These clients shouldn't be bound (def) in the codebase, as they may die, etc...

File: src/kafka/kafka_clients.clj [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+(comment
+
+  ;; generate record and send using producer
+  (def record (new ProducerRecord "demo" "Hello world!"))
+  (send-msg! producer record)
+
+  ;; consumer while loop to receive message
+  (read-msg! consumer)
+
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj:114]: An RCB should be demonstrating function use. This is missing numerous fns, and doesn't show any values from the fns

File: src/kafka/kafka_clients.clj [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+(defn build-producer-config
+  []
+  (let [producer-config (new Properties)]
+    (.putAll producer-config {"value.serializer"  StringSerializer
+                              "key.serializer"    StringSerializer
+                              "client.id"         (.getHostName (InetAddress/getLocalHost))
+                              "bootstrap.servers" "localhost:9092"
+                              "acks"              "all"})
+    producer-config))
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj]: For clarity, in your `let` forms only the final return value should be outside of the binding blocks `[]`

File: src/kafka/kafka_clients.clj [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+(defn build-consumer
+  [config]
+  (new KafkaConsumer config))
+
+(defn send-msg!
+  [producer record]
+  (.send producer record))
+
+(defn subscribe-topic-s!
```
Comment (leowei01) [inline src/kafka/kafka_clients.clj]: integrate into build-consumer

File: src/kafka/kafka_clients.clj [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+
+  ;; generate record
+  (do (def record1 (new ProducerRecord "demo" "Hello world!"))
+      record1)
+  #_=> object [org.apache.kafka.clients.producer.ProducerRecord
+               0x3fb54ba6
+               "ProducerRecord(topic=demo, partition=null, headers=RecordHeaders(headers = [], isReadOnly = false), key=null, value=Hello world!, timestamp=null)"]
+
+  ;; producer send record
+  (send-msg! producer record1)
```
Comment (leowei01) [inline src/kafka/kafka_clients.clj]: expected result should be a future

File: src/kafka/kafka_clients.clj [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+        "bootstrap.servers"  "localhost:9092",
+        "client.id"          "m-w1kx3xw0nr"}
+
+  ;; build producer
+  (do (def producer (build-producer producer-config)))
+  #_=> #'kafka.kafka-clients/producer
+
+  ;; build consumer
+  (do (def consumer (build-consumer consumer-config)))
+  #_=> #'kafka.kafka-clients/consumer
```
Comment (leowei01) [inline src/kafka/kafka_clients.clj]: expected result should be consumer object

File: src/kafka/kafka_clients.clj [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+                 "org.apache.kafka.clients.producer.internals.FutureRecordMetadata@de5318c"]
+
+  ;; consumer receive message
+  (read-msg! consumer)
+  #_=> ["Processed Value: Hello world!"]
+
+
+
+  ;; if the producer/consumer died, we should rebuild them
+  ;; if we want to change the config, we should rebuild the consumer and producer
```
Comment (leowei01) [inline src/kafka/kafka_clients.clj]: Here are some scenarios where I think atom can be used please leave your review if there is anything wrong.

File: src/kafka/kafka_clients.clj [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+  (atom {"group.id"           "My-Group"
+         "key.deserializer"   StringDeserializer
+         "value.deserializer" StringDeserializer
+         "client.id"          (.getHostName (InetAddress/getLocalHost))
+         "bootstrap.servers"  "localhost:9092"}))
+
+(defn build-producer-atom
+  [config]
+  (let [config (deref config)]
+    (atom (new KafkaProducer config))))
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj]: atoms (state) should be a top level `def`

File: src/kafka/kafka_clients.clj [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+
+(defn build-consumer-atom
+  [config]
+  (let [config   (deref config)
+        consumer (new KafkaConsumer config)]
+    (.subscribe consumer ["demo"])
+    (atom consumer)))
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj]: fns are unnecessary to build atoms. You should have fns to manipulate atoms, not generate them

File: src/kafka/kafka_clients.clj [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+    (.send producer record)))
+
+(defn read-msg!
+  [consumer]
+  (let [consumer (deref consumer)
+        record-s (.poll consumer (Duration/ofMillis 100))]
+    (.commitAsync consumer)
+    (mapv (fn [record]
+            (str "Processed Value: " (.value record)))
+          record-s)))
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj]: No code outside of the let block!!!!

File: src/kafka/kafka_clients.clj [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+  [config]
+  (let [config   (deref config)
+        consumer (new KafkaConsumer config)]
+    (.subscribe consumer ["demo"])
+    (atom consumer)))
+
+(defn send-msg!
+  [producer record]
+  (let [producer (deref producer)]
+    (.send producer record)))
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj]: No code outside of the let block!!!

File: src/kafka/kafka_clients.clj [hunk 19]
```diff
@@ -0,0 +1,170 @@
...
+        record-s (.poll consumer (Duration/ofMillis 100))]
+    (.commitAsync consumer)
+    (mapv (fn [record]
+            (str "Processed Value: " (.value record)))
+          record-s)))
+
+(defn build-producer
+  [config]
+  (let [config (deref config)]
+    (new KafkaProducer config)))
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj]: No code outside of the let block!!!!

File: src/kafka/kafka_clients.clj [hunk 20]
```diff
@@ -0,0 +1,170 @@
...
+
+(defn build-consumer
+  [config]
+  (let [config   (deref config)
+        consumer (new KafkaConsumer config)]
+    (.subscribe consumer ["demo"])
+    consumer))
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj:33]: No code outside of the let block!!!!

File: src/kafka/kafka_clients.clj [hunk 21]
```diff
@@ -0,0 +1,170 @@
...
+        "bootstrap.servers" "localhost:9092",
+        "acks"              "all"}
+
+  ;; modify consumer config
+  (swap! consumer-config assoc "client.id" "id")
+  #_=> {"group.id"           "My-Group",
+        "key.deserializer"   org.apache.kafka.common.serialization.StringDeserializer,
+        "value.deserializer" org.apache.kafka.common.serialization.StringDeserializer,
+        "client.id"          "id",
+        "bootstrap.servers"  "localhost:9092"}
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj]: You don't need to change the configurations. You _do_ need to the change (swap) the clients, and there should be fns to do so

File: src/kafka/kafka_clients.clj (hunk 20 above)
Comment (leowei01) [reply src/kafka/kafka_clients.clj:33]: Here I want to let the consumer subscribe topic and return the consumer, how could I move them to the let binding? The subscribe returns nil, should I use `(let [subscribe-result (.subscribe consumer ["demo"])])` ?

File: src/kafka/kafka_clients.clj (hunk 17 above)
Comment (leowei01) [reply src/kafka/kafka_clients.clj]: Here I'm going commit and return the records, but the commit will return nil, should I use `(let [commit-result (.commitAsync consumer)])` to move it into let binding?

File: src/kafka/kafka_clients.clj (hunk 21 above)
Comment (leowei01) [reply src/kafka/kafka_clients.clj]: If the client died, "change clients" means rebuild a client right? There should be fns take the client as parameter and rebuild the client, just like `(swap! client rebuild-client)`?

File: src/kafka/kafka_clients.clj (hunk 16 above)
Comment (leowei01) [reply src/kafka/kafka_clients.clj]: The atom should be `def` in the top level, just like `(def producer (atom (build-producer config)))`?

File: src/kafka/kafka_clients.clj (hunk 17 above)
Comment (lukaszhall) [reply src/kafka/kafka_clients.clj]: yes, but that suggested name is poor, and we have an idiomatic naming convention for unused bindings/symbols

File: src/kafka/kafka_clients.clj (hunk 20 above)
Comment (lukaszhall) [reply src/kafka/kafka_clients.clj:33]: See previous comment

File: src/kafka/kafka_clients.clj [hunk 22]
```diff
@@ -0,0 +1,170 @@
...
+
+
+  ;; modify producer with new config (this build-producer will return a Kafka producer instead of atom)
+  (reset! producer (build-producer producer-config))
+  #_=> #_#object[org.apache.kafka.clients.producer.KafkaProducer
+                 0x576e0952
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj]: you should not be directly manipulating the atoms in the RCB. You should have fns to do this for you.

File: src/kafka/kafka_clients.clj (hunk 21 above)
Comment (lukaszhall) [reply src/kafka/kafka_clients.clj]: Not quite. You are still framing this from the perspective of the implementation, which is backwards. Consider this from the perspective of the usage patterns.

e.g., generically:
```
(def some-atom .. )

(defn build-a-disposable-thing
   [config]
   ,,,)

(defn setup-some-atom-with-the-disposable-thing!
  [new-thing]
  (swap! some-atom ...))
```

File: src/kafka/kafka_clients.clj (hunk 22 above)
Comment (lukaszhall) [reply src/kafka/kafka_clients.clj]: see above comment for example

File: src/kafka/kafka_clients.clj [hunk 23]
```diff
@@ -0,0 +1,149 @@
+(ns kafka.kafka-clients
+  (:import (java.net InetAddress)
+           (org.apache.kafka.clients.producer KafkaProducer ProducerRecord)
+           (org.apache.kafka.common.serialization StringSerializer StringDeserializer)
+           (org.apache.kafka.clients.consumer KafkaConsumer)
+           (java.time Duration)))
+
+(def producer-config-atom
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj]: Avoid hungarian notation. i.e., `-atom`

File: src/kafka/kafka_clients.clj [hunk 24]
```diff
@@ -0,0 +1,149 @@
...
+         "acks"              "all"}))
+
+(def consumer-config-atom
+  (atom {"group.id"           "My-Group"
+         "key.deserializer"   StringDeserializer
+         "value.deserializer" StringDeserializer
+         "client.id"          (.getHostName (InetAddress/getLocalHost))
+         "bootstrap.servers"  "localhost:9092"}))
+
+(defn build-producer
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj:24]: This function is non-idempotent but is name idempotently. Make it idempotent

File: src/kafka/kafka_clients.clj [hunk 25]
```diff
@@ -0,0 +1,149 @@
...
+  []
+  (new KafkaProducer @producer-config-atom))
+
+(defn build-consumer
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj:28]: This function is non-idempotent but is named idempotently. Make it idempotent

File: src/kafka/kafka_clients.clj [hunk 26]
```diff
@@ -0,0 +1,149 @@
...
+  []
+  (let [consumer (new KafkaConsumer @consumer-config-atom)
+        _        (.subscribe consumer ["demo"])]
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj]: The topic is part of the configuration used to build the consumer

File: src/kafka/kafka_clients.clj [hunk 27]
```diff
@@ -0,0 +1,149 @@
...
+    consumer))
+
+(def producer-atom (atom (build-producer)))
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj]: Poor name. avoid Hungarian notation.

File: src/kafka/kafka_clients.clj [hunk 28]
```diff
@@ -0,0 +1,149 @@
...
+
+(def consumer-atom (atom (build-consumer)))
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj]: Poor name. avoid Hungarian notation.

File: src/kafka/kafka_clients.clj [hunk 29]
```diff
@@ -0,0 +1,149 @@
...
+
+(defn setup-producer-atom-with-new-producer!
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj]: Poor name

File: src/kafka/kafka_clients.clj [hunk 30]
```diff
@@ -0,0 +1,149 @@
...
+  [new-producer]
+  (reset! producer-atom new-producer))
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj]: `reset!` is not a safe operation

File: src/kafka/kafka_clients.clj [hunk 31]
```diff
@@ -0,0 +1,149 @@
...
+
+(defn setup-consumer-atom-with-new-consumer!
+  [new-consumer]
+  (reset! consumer-atom new-consumer))
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj]: `reset!` is not a safe operation

File: src/kafka/kafka_clients.clj [hunk 32]
```diff
@@ -0,0 +1,149 @@
...
+
+(defn send-msg!
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj:43]: make this idempotent

File: src/kafka/kafka_clients.clj [hunk 33]
```diff
@@ -0,0 +1,149 @@
...
+  [record]
+  (.send @producer-atom record))
+
+(defn read-msg!
+  []
+  (let [consumer @consumer-atom
+        record-s (.poll consumer (Duration/ofMillis 100))
+        value-s  (mapv (fn [record]
+                         (str "Processed Value: " (.value record)))
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj]: Why pollute the value? "Processed Value: " looks like an artifact from an old print/println.

File: src/kafka/kafka_clients.clj [hunk 34]
```diff
@@ -0,0 +1,149 @@
...
+
+  ;; generate record
+  (do (def record (new ProducerRecord "demo" "Hello world!"))
+      record)
+  #_=> #_#object[org.apache.kafka.clients.producer.ProducerRecord
+                 0x3fb54ba6
+                 "ProducerRecord(topic=demo, partition=null, headers=RecordHeaders(headers = [], isReadOnly = false), key=null, value=Hello world!, timestamp=null)"]
+
+  ;; producer send record
+  (send-msg! record)
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj]: Problem abstraction/decomposition issue. Seems like you would want this function to accept a string?

File: src/kafka/kafka_clients.clj [hunk 35]
```diff
@@ -0,0 +1,149 @@
...
+  ;; consumer receive message
+  (read-msg!)
+  #_=> ["Processed Value: Hello world!"]
+
+  ;; build new producer
+  (do (def new-producer (build-producer))
+      new-producer)
+  #_=> #_#object[org.apache.kafka.clients.producer.KafkaProducer
+                 0x56863721
+                 "org.apache.kafka.clients.producer.KafkaProducer@56863721"]
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj]: This demonstration is likely unnecessary, as `build-producer` is demonstrated earlier, and the atom is also already demonstrated.  (The following function that demonstrated updating the atom is sufficient)

File: src/kafka/kafka_clients.clj [hunk 36]
```diff
@@ -0,0 +1,149 @@
...
+
+  ;; build new client
+  (do (def new-consumer (build-consumer))
+      new-consumer)
+  #_=> #_#object[org.apache.kafka.clients.consumer.KafkaConsumer
+                 0x1129a6bf
+                 "org.apache.kafka.clients.consumer.KafkaConsumer@1129a6bf"]
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj]: This demonstration is likely unnecessary, as `build-consumer` is demonstrated earlier, and the atom is also already demonstrated.

File: src/kafka/kafka_clients.clj [hunk 37]
```diff
@@ -0,0 +1,149 @@
...
+  (:import (java.net InetAddress)
+           (org.apache.kafka.clients.producer KafkaProducer ProducerRecord)
+           (org.apache.kafka.common.serialization StringSerializer StringDeserializer)
+           (org.apache.kafka.clients.consumer KafkaConsumer)
+           (java.time Duration)))
+
+(def producer-config-atom
+  (atom {"value.serializer"  StringSerializer
+         "key.serializer"    StringSerializer
+         "client.id"         (.getHostName (InetAddress/getLocalHost))
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj]: keep this a just data - no fn calls

File: src/kafka/kafka_clients.clj (hunk 31 above)
Comment (leowei01) [reply src/kafka/kafka_clients.clj]: The function `setup-consumer-atom-with-new-consumer!` already has the new consumer to update atom. If I am going to use `swap!` here, the `fn` in `(swap! consumer fn)` will take consumer as input and update it with a new consumer, how about this way:
```
(defn fn
   [consumer new-consumer]
   new-consumer)

(swap! consumer fn new-consumer)
```

File: src/kafka/kafka_clients.clj (hunk 27 above)
Comment (leowei01) [reply src/kafka/kafka_clients.clj]: How about just name it `producer`? （I named it `producer-atom` to make it easy to distinguish between producer and producer atom)

File: src/kafka/kafka_clients.clj (hunk 29 above)
Comment (leowei01) [reply src/kafka/kafka_clients.clj]: how about `update-producer!`?

File: src/kafka/kafka_clients.clj (hunk 31 above)
Comment (leowei01) [reply src/kafka/kafka_clients.clj]: It seems like the `new-consumer` is not a good name, do you have any suggestions about `replacement-consumer`?

File: src/kafka/kafka_clients.clj [hunk 38]
```diff
@@ -0,0 +1,132 @@
...
+(defn build-producer
+  [configs]
+  (let [producer-config (@configs :producer-config)]
+    (new KafkaProducer producer-config)))
+
+(defn build-consumer
+  [configs]
+  (let [consumer-config (@configs :consumer-config)
+        consumer        (new KafkaConsumer consumer-config)
+        _               (.subscribe consumer ["demo"])]
```
Comment (leowei01) [inline src/kafka/kafka_clients.clj]: I don't know how to build consumer with the configuration including topics

File: src/kafka/kafka_clients.clj [hunk 39]
```diff
@@ -0,0 +1,132 @@
...
+                           "client.id"          "m-w1kx3xw0nr"
+                           "bootstrap.servers"  "localhost:9092"}}))
+
+(defn build-producer
+  [configs]
+  (let [producer-config (@configs :producer-config)]
+    (new KafkaProducer producer-config)))
+
+(defn build-consumer
+  [configs]
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj]: This is named as configs, but based on code below an atom is being passed?

File: src/kafka/kafka_clients.clj (hunk 38 above)
Comment (lukaszhall) [reply src/kafka/kafka_clients.clj]: You control the data structure(s) for `configs`... you can make this whatever want them to be.

File: src/kafka/kafka_clients.clj [hunk 40]
```diff
@@ -0,0 +1,132 @@
...
+  (let [consumer-config (@configs :consumer-config)
+        consumer        (new KafkaConsumer consumer-config)
+        _               (.subscribe consumer ["demo"])]
+    consumer))
+
+(def clients
+  (atom {:producer (build-producer configs)
+         :consumer (build-consumer configs)}))
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj]: Keep atom/state definitions at the top/head of the file

File: src/kafka/kafka_clients.clj [hunk 41]
```diff
@@ -0,0 +1,132 @@
...
+
+(defn send-msg!
+  [clients topic value]
+  (let [producer (@clients :producer)
+        record   (new ProducerRecord topic value)]
+    (.send producer record)))
+
+(defn read-msg!
+  [clients]
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj]: named as client, but passing an atom

File: src/kafka/kafka_clients.clj [hunk 42]
```diff
@@ -0,0 +1,132 @@
...
+  (let [consumer (@clients :consumer)
+        record-s (.poll consumer (Duration/ofMillis 100))
+        value-s  (mapv (fn [record]
+                         (.value record))
+                       record-s)
+        _        (.commitAsync consumer)]
+    value-s))
+
+(defn setup-clients-with-replacement-producer!
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj]: clean up name

File: src/kafka/kafka_clients.clj [hunk 43]
```diff
@@ -0,0 +1,132 @@
...
+  [clients replacement-producer]
+  (swap! clients assoc :producer replacement-producer))
+
+(defn setup-clients-with-replacement-consumer!
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj]: clean up name

File: src/kafka/kafka_clients.clj (hunk 39 above)
Comment (leowei01) [reply src/kafka/kafka_clients.clj]: But we should avoid hungarian notation (i.e., configs-atom) right?
Should I pass the consumer config here 
(call this function with `(build-consumer (@configs :consumer-config))`) ?

File: src/kafka/kafka_clients.clj (hunk 43 above)
Comment (leowei01) [reply src/kafka/kafka_clients.clj]: how about `update-consumer-client!`

File: src/kafka/kafka_clients.clj (hunk 41 above)
Comment (leowei01) [reply src/kafka/kafka_clients.clj]: But we should avoid hungarian notation (i.e., clients-atom) right?
Should I pass the consumer client here 
(call this function with `(read-msg! (@clients :consumer))`) ?

File: src/kafka/kafka_clients.clj (hunk 40 above)
Comment (leowei01) [reply src/kafka/kafka_clients.clj]: Here I need the `build-producer` and `build-consumer` functions to build atom, so the functions' definition must be on top of the atom definition...

Comment (lukaszhall) [reply src/kafka/kafka_clients.clj]: see `declare`. Also, nothing prevents these from starting as `nil`

File: src/kafka/kafka_clients.clj (hunk 41 above)
Comment (lukaszhall) [reply src/kafka/kafka_clients.clj]: Do you think we should be passing state around, or data?

Comment (leowei01) [reply src/kafka/kafka_clients.clj]: Should pass data

File: src/kafka/kafka_clients.clj [hunk 44]
```diff
@@ -0,0 +1,132 @@
...
+  [clients replacement-consumer]
+  (swap! clients assoc :consumer replacement-consumer))
+
+
+
+
+
+(comment
+
+  configs
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj]: Don't need to see the atom, need to see just the value

File: src/kafka/kafka_clients.clj [hunk 45]
```diff
@@ -0,0 +1,132 @@
...
+                                             "client.id"         "m-w1kx3xw0nr",
+                                             "bootstrap.servers" "localhost:9092",
+                                             "acks"              "all"},
+                           :consumer-config {"group.id"           "My-Group",
+                                             "key.deserializer"   org.apache.kafka.common.serialization.StringDeserializer,
+                                             "value.deserializer" org.apache.kafka.common.serialization.StringDeserializer,
+                                             "client.id"          "m-w1kx3xw0nr",
+                                             "bootstrap.servers"  "localhost:9092"}}}]
+
+  clients
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj]: Don't need to see the atom, need to see just the value

File: src/kafka/kafka_clients.clj [hunk 46]
```diff
@@ -0,0 +1,125 @@
...
+                          "bootstrap.servers" "localhost:9092",
+                          "acks"              "all"},
+        :consumer-config {"group.id"           "My-Group",
+                          "key.deserializer"   org.apache.kafka.common.serialization.StringDeserializer,
+                          "value.deserializer" org.apache.kafka.common.serialization.StringDeserializer,
+                          "client.id"          "m-w1kx3xw0nr",
+                          "bootstrap.servers"  "localhost:9092",
+                          "topic"              "demo"}}
+
+  @clients
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj:74]: These would typically be nil until initialized.

File: src/kafka/kafka_clients.clj [hunk 47]
```diff
@@ -0,0 +1,118 @@
...
+                           "client.id"         "m-w1kx3xw0nr"
+                           "bootstrap.servers" "localhost:9092"
+                           "acks"              "all"}
+         :consumer-config {"group.id"           "My-Group"
+                           "key.deserializer"   StringDeserializer
+                           "value.deserializer" StringDeserializer
+                           "client.id"          "m-w1kx3xw0nr"
+                           "bootstrap.servers"  "localhost:9092"
+                           "topic"              "demo"}}))
+
```
Comment (lukaszhall) [inline src/kafka/kafka_clients.clj:19]: (def counter (atom 0))

**Pull request merged**